kart/diff_structs.py

```diff
@@ -432,117 +438,96 @@ def type_counts(self):
-    contents in memory, which may be expensive.
+    To consume the iterator into memory and turn the LazyDeltaDiff into a DeltaDiff, call `resolve()`
     """
 
-    child_type = Delta
+    _wrapped_iter: Iterator[Delta]
 
-    def __init__(self, initial_contents=()):
-        # An iterator over keys and Delta objects, which is consumed lazily
-        self._lazy_initial_contents: Iterator[tuple[str, Delta]] | None = None
-        if isinstance(initial_contents, (dict, UserDict)):
-            super().__init__(initial_contents)
+    def __init__(self, initial_contents: Iterable[Delta] = ()):
+        wrapped_iter = iter(initial_contents)
+        try:
+            first_item = next(wrapped_iter)
+        except StopIteration:
+            self._wrapped_iter = iter(())
+            self._bool = False
         else:
-            if isinstance(initial_contents, Iterator):
-                self._lazy_initial_contents = (
-                    (delta.key, delta) for delta in initial_contents
-                )
-                initial_contents = ()
-            super().__init__((delta.key, delta) for delta in initial_contents)
-
-    def __getitem__(self, key):
-        if key in self.data:
-            return self.data[key]
-        self._evaluate_lazy_initial_contents()
-        return self.data[key]
-
-    def __setitem__(self, key, delta):
-        if key != delta.key:
-            raise ValueError("Delta must be added at the appropriate key")
-        super().__setitem__(key, delta)
-
-    def _evaluate_lazy_initial_contents(self):
-        if self._lazy_initial_contents is None:
-            return
-        # Consume the generator to populate the DeltaDiff.
-        self.update(self._lazy_initial_contents)
-        self._lazy_initial_contents = None
+            self._wrapped_iter = chain((first_item,), wrapped_iter)
+            self._bool = True
+        self._consumed = False
 
     def __bool__(self):
-        result = bool(self.data)
-        if (not result) and self._lazy_initial_contents:
-            # If the DeltaDiff is empty, but has lazy initial contents, evaluate the first item to check booleanness.
-            try:
-                k, v = next(self._lazy_initial_contents)
-            except StopIteration:
-                return False
-            else:
-                # remember this result
-                self.data[k] = v
-                return True
-        return result
+        return self._bool
 
-    def __len__(self):
-        self._evaluate_lazy_initial_contents()
-        return super().__len__()
+    def __add__(self, other):
+        resolved = self.resolve()
+        resolved += other
+        return resolved
 
-    def items(self):
-        self._evaluate_lazy_initial_contents()
-        return super().items()
+    def _check_not_consumed(self):
+        if self._consumed:
+            raise ValueError("LazyDeltaDiff has already been consumed")
```

Comment:
Match the error if you muck with dicts while iterating?

```suggestion
            raise RuntimeError("LazyDeltaDiff has already been consumed")
```